superset/common/query_context_processor.py

```diff
@@ -307,6 +324,24 @@ def _get_timestamp_format(
 
         return df
 
+    @staticmethod
+    def get_time_grain(
+        query_context: QueryContext, query_object: QueryObject
+    ) -> Optional[Any]:
+        if query_context.form_data:
+            return query_context.form_data.get("time_grain_sqla")
+
+        if (
+            query_object.columns
+            and len(query_object.columns) > 0
+            and isinstance(query_object.columns[0], dict)
+        ):
+            # If the time grain is in the columns it will be the first one
+            # and it will be of AdhocColumn type
+            return query_object.columns[0].get("timeGrain")
+
+        return query_object.extras.get("time_grain_sqla")
```

Comment:
The joys of Superset and its N different ways to store/retrieve the same information 🤦